src/intptrcast.rs

```diff
@@ -34,24 +40,36 @@ impl GlobalStateInner {
         GlobalStateInner {
             int_to_ptr_map: Vec::default(),
             base_addr: FxHashMap::default(),
+            exposed: FxHashSet::default(),
             next_base_addr: STACK_ADDR,
+            permissive_provenance: config.permissive_provenance,
             strict_provenance: config.strict_provenance,
         }
     }
 }
 
 impl<'mir, 'tcx> GlobalStateInner {
-    pub fn ptr_from_addr(addr: u64, ecx: &MiriEvalContext<'mir, 'tcx>) -> Pointer<Option<Tag>> {
-        trace!("Casting 0x{:x} to a pointer", addr);
+    // Returns the `AllocId` that corresponds to the specified addr,
+    // or `None` if the addr is out of bounds
+    fn alloc_id_from_addr(ecx: &MiriEvalContext<'mir, 'tcx>, addr: u64) -> Option<AllocId> {
         let global_state = ecx.machine.intptrcast.borrow();
 
-        if global_state.strict_provenance {
-            return Pointer::new(None, Size::from_bytes(addr));
+        if addr == 0 {
+            return None;
         }
```

Comment:
That errors because of this code here:

https://github.com/rust-lang/rust/blob/625e4dd13a3abd0cc59807af66c3c4cd63440852/compiler/rustc_const_eval/src/interpret/memory.rs#L1124

I can move the check to `ptr_from_addr`, but https://llvm.org/docs/GetElementPtr.html#can-i-cast-an-object-s-address-to-integer-and-add-it-to-null implies that LLVM will reject this. Now LLVM doesn't support the semantics in this PR, but it seemed important if there was an assert for it, so I think we need this special-cased somewhere.

Reply:
The search below will never find anything for the null pointer anyway, will it? So the return value should be `None` either way.

Reply:
Hmm yeah these semantics are actually subtly wrong... I'll fix it once I get around to fixing this part of the code.